Source/PINCacheKeyObserverManager.m

```diff
@@ -0,0 +1,147 @@
+    if (self) {
+        _cache = cache;
+        _keysToObserversMapping = [[NSMutableDictionary alloc] init];
+        NSString *queueName = [NSString stringWithFormat:@"%@.key_observer_manager", PINDiskCachePrefix];
+        _queue = dispatch_queue_create([queueName UTF8String], DISPATCH_QUEUE_SERIAL);
+        _observersNotificationCenter = [[NSNotificationCenter alloc] init];
+    }
+    return self;
+}
+
+#pragma mark - Public methods
+
+- (void)updatedKey:(NSString *)key withValue:(id)value
+{
+    dispatch_async(_queue, ^{
+        if (key.length == 0 || ![self isTrackingKey:key]) {
+            return;
+        }
+        
+        NSDictionary *userInfo = [self userInfoForKey:key withValue:value];
+        [_observersNotificationCenter postNotificationName:[self notificationNameForKey:key]
+                                                    object:nil
+                                                  userInfo:userInfo];
+    });
+}
+
+- (void)deletedValueForKey:(NSString *)key
+{
+    dispatch_async(_queue, ^{
+        if (key.length == 0 || ![self isTrackingKey:key]) {
+            return;
+        }
+        
+        NSDictionary *userInfo = [self userInfoForKey:key withValue:nil];
+        [_observersNotificationCenter postNotificationName:[self notificationNameForKey:key]
+                                                    object:nil
+                                                  userInfo:userInfo];
+    });
+}
+
+- (void)deletedAllValues
+{
+    for (NSString *key in [_keysToObserversMapping allKeys]) {
+        [self deletedValueForKey:key];
+    }
+}
+
+#pragma mark - PINCacheKeyObserving
+
+- (void)addObserver:(id)observer selector:(SEL)selector forKey:(NSString *)key
+{
+    if (key.length == 0) {
+        return;
+    }
+    
+    dispatch_async(_queue, ^{
+        [_observersNotificationCenter addObserver:observer
+                                         selector:selector
+                                             name:[self notificationNameForKey:key]
+                                           object:nil];
+        
+        NSMutableSet<id> *observers = _keysToObserversMapping[key];
+        if (observers == nil) {
+            observers = [[NSMutableSet alloc] init];
+            _keysToObserversMapping[key] = observers;
+        }
+        [observers addObject:observer];
+        
+    });
+}
+
+- (void)removeObserver:(id)observer forKey:(NSString *)key
+{
+    if (key.length == 0) {
+        return;
+    }
+    
+    dispatch_async(_queue, ^{
+        [_observersNotificationCenter removeObserver:observer
+                                                name:[self notificationNameForKey:key]
+                                              object:nil];
+        
+        NSMutableSet<id> *observers = _keysToObserversMapping[key];
+        [observers removeObject:observer];
+    });
+}
+
+#pragma mark - Private
+
+- (BOOL)isTrackingKey:(NSString *)key
+{
+    NSSet<id> *set = _keysToObserversMapping[key];
```

Comment:
I don't think `_keysToObserversMapping` is safe to access outside of the serial queue?